src/overload_ifelse.jl

```diff
@@ -0,0 +1,41 @@
+const SimpleTracer = Union{ConnectivityTracer,GradientTracer,HessianTracer}
+
+function Base.ifelse(::T, x, y) where {T<:SimpleTracer}
+    size(x) != size(y) &&
+        throw(DimensionMismatch("Outputs sizes of `ifelse` arguments don't match in size."))
+    return output_union(x, y)
+end
+
+## output union on scalar outputs
+output_union(tx::C, ty::C) where {C<:ConnectivityTracer} = C(inputs(tx) ∪ inputs(ty))
+output_union(tx::G, ty::G) where {G<:GradientTracer} = G(gradient(tx) ∪ gradient(ty))
+function output_union(tx::H, ty::H) where {H<:HessianTracer}
+    return H(gradient(tx) ∪ gradient(ty), hessian(tx) ∪ hessian(ty))
+end
+
+output_union(tx::T, y) where {T<:SimpleTracer} = tx
+output_union(x, ty::T) where {T<:SimpleTracer} = ty
+
+## output union on AbstractArray outputs
+function output_union(tx::AbstractArray{T}, ty::AbstractArray{T}) where {T<:SimpleTracer}
+    return output_union.(tx, ty)
+end
+function output_union(tx::AbstractArray{T}, y::AbstractArray) where {T<:SimpleTracer}
+    return tx
+end
+function output_union(x::AbstractArray, ty::AbstractArray{T}) where {T<:SimpleTracer}
+    return ty
+end
+
+## Overload comparisons
+# We don't return empty tracers since the resulting Bool could be used as a Number. 
+for fn in (:isequal, :isapprox, :isless, :(==), :(<), :(>), :(<=), :(>=))
+    @eval Base.$fn(tx::C, ty::C) where {C<:ConnectivityTracer} = C(inputs(tx) ∪ inputs(ty))
+    @eval Base.$fn(tx::G, ty::G) where {G<:GradientTracer} = G(gradient(tx) ∪ gradient(ty))
+    @eval function Base.$fn(tx::H, ty::H) where {H<:HessianTracer}
+        return H(gradient(tx) ∪ gradient(ty), hessian(tx) ∪ hessian(ty))
```

Comment:
we don't need the other terms of hessian tracing (outer products of gradients) here?

Reply:
I've thought about this a bit and now suspect it's actually the opposite.
`>` and friends are just regular 2-to-1 operators that are return constant 0 or 1 over their input domain. Their derivatives are therefore zero everywhere.

As a reminder, `hessian_tracer_2_to_1` is defined as
``` 
    grad = empty(G)
    hess = empty(H)
    if !is_firstder_arg1_zero
        grad = union(grad, gradient(a)) # TODO: use union!
        union!(hess, hessian(a))
    end
    if !is_firstder_arg2_zero
        grad = union(grad, gradient(b)) # TODO: use union!
        union!(hess, hessian(b))
    end
    if !is_seconder_arg1_zero
        union!(hess, gradient(a) × gradient(a))
    end
    if !is_seconder_arg2_zero
        union!(hess, gradient(b) × gradient(b))
    end
    if !is_crossder_zero
        union!(hess, (gradient(a) × gradient(b)) ∪ (gradient(b) × gradient(a)))
    end
    return T(grad, hess)
```

The correct implementation therefore is 
```Julia
@eval function Base.$fn(tx::H, ty::H) where {G,H,T<:HessianTracer{G,H}}
    return T(gradient(tx) ∪ gradient(ty), empty(H))
end
```